File: python/euler/bin/p0077.py
# project euler: problem 77

from collections.abc import Iterator

from euler.lib.prime import prime_generator


def plst_generator() -> Iterator[list[int]]:
    p_gen = prime_generator()
    plst = []
    while True:
        plst.append(next(p_gen))
        yield plst


def compute(boundary: int) -> str:
    plst_gen = plst_generator()
    while True:
        plst = next(plst_gen)
        tbl = [0] * (len(plst) + 1)
        tbl[0] = 1
        for i in plst:
            # Skip when i >= len(tbl)
            # This is because the prime number 'i' is larger than the sum of primes we now assume
            for j in range(i, len(tbl)):
                tbl[j] += tbl[j - i]

        # over the boundary
        if tbl[-1] > boundary:
            break

    return str(len(plst))


def solve() -> str:
    return compute(5_000)

File: typescript/euler/bin/p0077.ts
// project euler: problem 77

import { primeGenerator } from "../lib/primes.ts";
import { range } from "../lib/util.ts";

const plstGenerator = function* (): Generator<number[], void, unknown> {
  const p_gen = primeGenerator();
  const plst: number[] = [];
  while (true) {
    plst.push(p_gen.next().value as number);
    yield plst;
  }
};

export const compute = (boundary: number): string => {
  const plst_gen = plstGenerator();
  let plst: number[] = [];

  while (true) {
    plst = plst_gen.next().value as number[];
    const tbl: number[] = new Array(plst.length + 1).fill(0);
    tbl[0] = 1;
    for (const i of plst) {
      for (const j of range(i, tbl.length)) {
        tbl[j] += tbl[j - i];
      }
    }

    if (tbl.at(-1)! > boundary) {
      break;
    }
  }

  return String(plst.length);
};

export const solve = (): string => compute(5_000);
